### ISSUE TITLE
exp push: studio integration

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#9074]
To make `exp push/pull` more useful, users need to be able to see and manage the exp refs from within Studio (since those exp refs are ignored by GitHub and other Git providers). One of the challenges is that there is not even any trigger by which the Git providers can alert Studio that new exp refs have been pushed.

To enable Studio support for pushed experiments, DVC needs to:

```[tasklist]
-  [x] Support for saving Studio token to global DVC config or somewhere else that's reasonably secure, persistent, and ideally globally available
- [x] #9163
-  [x] Return links to experiments in Studio so user knows exactly where to see what was pushed - sufficient to return link to project for now
- [x] https://github.com/iterative/dvc/pull/9161
- [x] #9227
- [x] https://github.com/iterative/dvc/pull/9248
- [ ] #9280
- [ ] #9257
- [x] https://github.com/iterative/dvc.org/issues/4440 command ref updates
```

Blocked/out of scope issues:

```[tasklist]
- [ ] https://github.com/iterative/dvc/issues/9265
- [ ] https://github.com/iterative/dvc/issues/9295
```

Related:
- https://github.com/iterative/studio/issues/5050
- https://github.com/iterative/studio/issues/5223
- https://github.com/iterative/vscode-dvc/issues/3077

**Comment by mattseddon** [comment]
>  Support for saving Studio token to global DVC config or somewhere else that's reasonably secure, persistent, and ideally globally available

Github CLI has `gh auth` set of commands. The most useful for me has been [gh auth login](https://cli.github.com/manual/gh_auth_login).

**Comment by dberenbaum** [comment]
Edited to add:

> - [ ] Enable live sharing of metrics and exp refs for queued experiments (see https://github.com/iterative/dvclive/issues/485)

**Comment by daavoo** [comment]
>  Return links to experiments in Studio so user knows exactly where to see what was pushed

What would this link to? Is it the default project view or something else? Afaik there is no real concept of a view for a single experiment in Studio

**Comment by dberenbaum** [comment]
> What would this link to? Is it the default project view or something else? Afaik there is no real concept of a view for a single experiment in Studio

I think it's discussed in https://github.com/iterative/studio/issues/5150.

> * When pushing refs, alert Studio that new ref is available so Studio knows to parse the new refs

See https://github.com/iterative/studio/issues/5273#issuecomment-1462105892. @skshetry Looks like we already have the API needed for this.

**Comment by skshetry** [comment]
>  Looks like we already have the API needed for this.

I'll leave that upto the Studio team to decide. :)
At the moment, what we need is a friendly ping to Studio to go and fetch new exp refs and parse. The existing API seems to only use the data from the payload, does not fetch or parse anything other than that.

**Comment by skshetry** [comment]
> Return links to experiments in Studio so user knows exactly where to see what was pushed

This may be a bit tricky with just `STUDIO_TOKEN` and `repo_url` to figure out as a user can have multiple views.

**Comment by skshetry** [comment]
Some updates here: I am waiting for Studio to parse and show the experiment refs. We already have the functionality to notify Studio of the refs. The API endpoint does refresh and tries to parse new commits.

>  Support for saving Studio token to global DVC config or somewhere else that's reasonably secure, persistent, and ideally globally available

I am thinking of implementing support in dvc config for token and calling this item done here for now. Secure storage implementation can be done in the future.

> Return links to experiments in Studio so user knows exactly where to see what was pushed

Looks like this was deprioritized. So I don't see it happening any time soon.

---

I have 2 questions regarding the workflow:

1. If we did set a config, will it always by default push refs to Studio? Or should it be a separate config?
2. Should there be a separate config to push or not push refs to the Studio?
   Eg: `--studio` and `--no-studio` perhaps ?

**Comment by dberenbaum** [comment]
Sounds good @skshetry, thanks!

> I have 2 questions regarding the workflow:
> 
>     1. If we did set a config, will it always by default push refs to Studio? Or should it be a separate config?

I think it should push by default. It's a lightweight ping, and if it fails we can make it not too alarming. I don't see much reason not to push in most cases. We already will need to figure out how to differentiate this from the live sharing to studio that's enabled currently by the env var, so let's not add more options than necessary yet.

>     2. Should there be a separate config to push or not push refs to the Studio?
>        Eg: `--studio` and `--no-studio` perhaps ?

Is it needed? Similar to above, is there a situation where it's harmful to send a ping to Studio if the token exists?

Should we return an error code on failures to ping Studio? My only worry would be not breaking CI jobs if the ping to Studio fails.

**Comment by dberenbaum** [comment]
I see we might end up trying to avoid having a token according to https://github.com/iterative/studio/issues/5050#issuecomment-1479601463:

> One is to optimize the flow of trigger. Currently, user need to have the studio token to make webhook trigger. We can remove the auth part and use existing ip/user rate limit as safeguards. If the refs sent by the dvc exists in git forge, then only we can trigger the parse.

In that case, I can see more why we may need some way to enable/disable sending to Studio. It might be fine to ping by default and quietly fail since it's a pretty lightweight request, but I'd agree we probably need some way to opt out of this behavior.

**Comment by dberenbaum** [comment]
@skshetry Added #9227

**Comment by dberenbaum** [comment]
@skshetry I got an error when testing:

```bash
$ dvc exp push -v --no-cache origin whole-hate
2023-03-24 14:35:20,111 DEBUG: v2.51.1.dev15+g1f80f2f2e, CPython 3.10.2 on macOS-13.2.1-arm64-arm-64bit
2023-03-24 14:35:20,111 DEBUG: command: /Users/dave/miniforge3/envs/dvc/bin/dvc exp push -v --no-cache origin whole-hate
2023-03-24 14:35:20,328 DEBUG: git push experiment ['refs/exps/ac/68ea47f0fec5c78ce259f014ff1a3f07634e76/whole-hate:refs/exps/ac/68ea47f0fec5c78ce259f014ff1a3f07634e76/whole-hate'] -> 'origin'
2023-03-24 14:35:22,331 DEBUG: pushing experiments to Studio (None)
2023-03-24 14:35:22,503 ERROR: unexpected error - 400 Client Error: Bad Request for url: https://studio.iterative.ai/webhook/dvc/
Traceback (most recent call last):
  File "/Users/dave/Code/dvc/dvc/cli/__init__.py", line 210, in main
    ret = cmd.do_run()
  File "/Users/dave/Code/dvc/dvc/cli/command.py", line 26, in do_run
    return self.run()
  File "/Users/dave/Code/dvc/dvc/commands/experiments/push.py", line 26, in run
    pushed_exps = self.repo.experiments.push(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 491, in push
    return push(self.repo, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/__init__.py", line 65, in wrapper
    return f(repo, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/scm_context.py", line 151, in run
    return method(repo, *args, **kw)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/push.py", line 88, in push
    _notify_studio([str(ref) for ref in refs], repo_url, token, url=studio_url)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/push.py", line 134, in _notify_studio
    resp.raise_for_status()
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://studio.iterative.ai/webhook/dvc/

2023-03-24 14:35:22,546 DEBUG: Removing '/Users/dave/Code/.Gh943d6HDmmx8bUE3sTnCi.tmp'
2023-03-24 14:35:22,546 DEBUG: Removing '/Users/dave/Code/.Gh943d6HDmmx8bUE3sTnCi.tmp'
2023-03-24 14:35:22,546 DEBUG: Removing '/Users/dave/Code/.Gh943d6HDmmx8bUE3sTnCi.tmp'
2023-03-24 14:35:22,546 DEBUG: Removing '/Users/dave/Code/lstm_seq2seq/.dvc/cache/.N2aaxGyYbtHAfF4YMxTSbT.tmp'
2023-03-24 14:35:26,567 DEBUG: Version info for developers:
DVC version: 2.51.1.dev15+g1f80f2f2e
------------------------------------
Platform: Python 3.10.2 on macOS-13.2.1-arm64-arm-64bit
Subprojects:
        dvc_data = 0.45.1
        dvc_objects = 0.21.1
        dvc_render = 0.3.1
        dvc_task = 0.2.0
        scmrepo = 0.1.17
Supports:
        azure (adlfs = 2022.9.1, knack = 0.9.0, azure-identity = 1.7.1),
        gdrive (pydrive2 = 1.15.0),
        gs (gcsfs = 2022.11.0),
        hdfs (fsspec = 2022.11.0+0.gacad158.dirty, pyarrow = 7.0.0),
        http (aiohttp = 3.8.1, aiohttp-retry = 2.8.3),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.8.3),
        oss (ossfs = 2021.8.0),
        s3 (s3fs = 2022.11.0+6.g804057f.dirty, boto3 = 1.24.59),
        ssh (sshfs = 2023.1.0),
        webdav (webdav4 = 0.9.4),
        webdavs (webdav4 = 0.9.4),
        webhdfs (fsspec = 2022.11.0+0.gacad158.dirty)
Cache types: reflink, hardlink, symlink
Cache directory: apfs on /dev/disk3s1s1
Caches: local
Remotes: s3
Workspace directory: apfs on /dev/disk3s1s1
Repo: dvc, git
Repo.site_cache_dir: /Library/Caches/dvc/repo/67c1f9dac32e24b2ac3b76c1b8c45117
```

**Comment by skshetry** [comment]
@dberenbaum, you need to pull main branch.

**Comment by skshetry** [comment]
FYI you may have to sync with main for this to work. We changed the endpoint.

**Comment by skshetry** [comment]
The only thing that's required for release is https://github.com/iterative/dvclive/issues/488, but it's not a blocker on dvc side.

**Comment by daavoo** [comment]
> The only thing that's required for the release is https://github.com/iterative/dvclive/issues/488, but it's not a blocker on dvc side.

For `dvc exp run`, when live updates are enabled, we should discuss if we expect dvc to also call `exp push` here:

https://github.com/iterative/dvc/blob/efc2d7ad3d2cedcd3d13dc8ddc767751a7819431/dvc/repo/experiments/executor/base.py#L641-L649

DVCLive should do the same order of calls when `save_dvc_exp=True` and people run with `python dvclive_script.py`

**Comment by daavoo** [comment]
> we should discuss if we expect dvc to also call `exp push` here:

We could also remove the `metrics` argument if we push the full exp ref on completion

**Comment by dberenbaum** [comment]
@skshetry Could you open a docs issue/PR for this work 🙏 ?

**Comment by dberenbaum** [comment]
@skshetry I added this item:

- [ ] Get Studio token through `gh login`-like interface (https://github.com/iterative/dvc/issues/9265).

Also, are you still planning on this item?

- [ ]  Return links to experiments in Studio so user knows exactly where to see what was pushed - sufficient to return link to project for now

Even if it's a project-level URL, I think it would be helpful to both indicate the ping was successful and make it easy to navigate to Studio to see the experiments.

**Comment by skshetry** [comment]
> Also, are you still planning on this item?
> 
> Even if it's a project-level URL, I think it would be helpful to both indicate the ping was successful and make it easy to navigate to Studio to see the experiments.

It can be implemented later on when Studio starts returning URL.

**Comment by dberenbaum** [comment]
@skshetry Can we do it now then since it looks like it was just implemented in https://github.com/iterative/studio/pull/5509?

**Comment by skshetry** [comment]
> For `dvc exp run`, when live updates are enabled, we should discuss if we expect dvc to also call `exp push` here:

I need your opinion here, @dberenbaum. Note that we have `DVC_EXP_AUTO_PUSH` envvar that does this already.

I'd prefer being explicit here with `dvc exp run --push`. I don't really like to have too much magic with just `STUDIO_TOKEN` set.

**Comment by dberenbaum** [comment]
> Note that we have `DVC_EXP_AUTO_PUSH` envvar that does this already.

> I don't really like to have too much magic with just `STUDIO_TOKEN` set.

Maybe the problem is that we are using `STUDIO_TOKEN` for everything. We did it for simplicity at the start, but it's always felt like an odd way to enable/disable sharing to me. Should we use `DVC_EXP_AUTO_PUSH` to enable/disable live sharing instead? It aligns with the discussion in #9221 to move away from checkpoints.

It could work like:
1. Set `DVC_EXP_AUTO_PUSH` to get automatic sharing.
2. While the experiment runs, each DVCLive step will push updates to Studio using the Studio token (either from env var or the config). 
3. At the end/when a commit happens, DVC will do `dvc exp push`.

**Comment by shcheklein** [comment]
Are there any other ways to control besides using env vars (personal bias - I never liked things like `export SOMETHING` or remember to do `SOMETHING=true command` - all looks like hacks to me, when you need to override something temporarily vs a proper interface.

**Comment by dberenbaum** [comment]
If we use `DVC_EXP_AUTO_PUSH`, I see no reason we can't create a DVC config option with the same name.

**Comment by shcheklein** [comment]
Can it be / should it be controlled on the code level as well? E.g. enabled by default is token is set + an option to opt out.?

**Comment by dberenbaum** [comment]
@shcheklein Can we agree on the schema and follow up on the exact workflow? Any thoughts on using both `STUDIO_TOKEN` and `DVC_EXP_AUTO_PUSH` to control this behavior? 

Using both of those options as described above unifies the live sharing and exp push workflow, and it gives flexibility to tweak the workflows. We can save the token and either share or not, and can change the default behavior (we could enable `DVC_EXP_AUTO_PUSH` by default when saving the token if we want). We can also set them from CLI, code, in environment variables, etc.

**Comment by shcheklein** [comment]
@dberenbaum could you please clarify the details a bit? I think it's hard to understand how all the things fit together. 

So, we have a token in the DVC config now (who is reading it and how do we pass it downstream in DVC). Af far as I understand it's also possible to set `STUDIO_TOKEN`, we do this in VS Code for example.

We don't use `DVC_EXP_AUTO_PUSH` yet (I guess push is enabled by default or controlled via code).

Anyways, I think we need to clarify the whole picture tbh here. It's hard to agree w/o knowing the details (or may be I don't know them enough.

**Comment by dberenbaum** [comment]
Here's a proposal:

* `STUDIO_TOKEN` is used to save the token and has no other effect.
* `DVC_EXP_AUTO_PUSH` is used to enable/disable automatic sharing of experiments. This would push exp refs whenever they are saved locally, and if there is a valid `STUDIO_TOKEN`, it would also share live updates.
* Both could be configured through different interfaces with a typical priority order: CLI/API flag -> environment variable -> local config -> project config -> user/global config -> system config.

**Comment by shcheklein** [comment]
okay, I see. So it's just names for configs, not env vars. So, the question here is to decide on names and see if those two are more or less enough and seem reasonable from the user perspective (assuming we can change and fine tune their exact behavior, etc), right?

>  and if there is a valid STUDIO_TOKEN, it would also share live updates.
> STUDIO_TOKEN is used to save the token and has no other effect.

It's a bit of a contradiction?

-----

A few things to think, from the top of my head:

- do we have other env / global configs similar to this? how do we name them (e.g. should be name all the options `DVC_something`, or other way around no `DVC_` prefix). E.g. should it be DVC_STUDIO_TOKEN? or ITERATIVE_STUDIO_TOKEN?
- can it happen that we would need `DVC_EXP_PUSH_MODE={auto,default,no}` (just making this up a bit- no need to spend too much on it- just feels a bit more future-proof?). (the answer here is probably no, since the name is confusing ... just in case)

**Comment by dberenbaum** [comment]
> > and if there is a valid STUDIO_TOKEN, it would also share live updates.
> > STUDIO_TOKEN is used to save the token and has no other effect.
> 
> It's a bit of a contradiction?

Maybe you took me a bit too literally 😄. Do you have confusion over how it will work in some scenario?

> * do we have other env / global configs similar to this? how do we name them (e.g. should be name all the options `DVC_something`, or other way around no `DVC_` prefix). E.g. should it be DVC_STUDIO_TOKEN? or ITERATIVE_STUDIO_TOKEN?

See https://github.com/iterative/dvc/blob/main/dvc/env.py. Agree that it should probably be `DVC_STUDIO_TOKEN`. We probably need a docs page listing all those.

There is also https://dvc.org/doc/dvclive/env.

@skshetry Maybe you have thoughts on sections/naming for both options in the config?

> * can it happen that we would need `DVC_EXP_PUSH_MODE={auto,default,no}` (just making this up a bit- no need to spend too much on it- just feels a bit more future-proof?). (the answer here is probably no, since the name is confusing ... just in case)

Not sure I get what each option there would mean. I think we could adjust the "default" behavior via the UX in #9265. For example:

```
$ dvc studio login
Redirecting you to studio.iterative.ai...
Token saved to `~/.config/dvc/config`
Would you like to enable automatic experiment sharing? (Y/n)
```

In other words, rather than have an option like `exp_auto_push` that defaults to true when unset, we can stick to typical convention (unset means false) but adjust default behavior through the UX.

**Comment by dberenbaum** [comment]
@skshetry Let's wrap this up by:
* Merging #9257 (the feature flag is already removed in Studio)
* Addressing the cmd ref for https://github.com/iterative/dvc.org/issues/4440

**Comment by omesser** [comment]
@dberenbaum can you please clarify where did we land on with the experience? I'm a bit worried we're adding friction to the user for the happy flow

**Comment by dberenbaum** [comment]
@omesser There's a lot of related discussion in #9295. For now, here's how it will work:

* Setting `STUDIO_TOKEN` will enable dvclive sharing. 
* If either `STUDIO_TOKEN` or the dvc config option `studio.token` are set, `dvc exp push` will ping Studio in addition to pushing to the git and/or dvc remotes.

Nothing else is planned for now. We discussed auto-pushing exp refs and the corresponding dvc cached outputs, but after the discussion in #9295, I think we should wait. We have taken some steps to make live updates feel connected to the rest of the experiments workflow ([like replacing the live row with the exp ref row in Studio](https://github.com/iterative/studio/issues/5453)), and I no longer feel they seem too disconnected. Live sharing is a different scenario than `dvc exp push` where it's unclear whether you want to push the exp ref or dvc cache automatically, and it makes explaining dvclive sharing more complicated (it pushes some lightweight stuff to Studio in realtime, and then pushes everything to the Git remote and DVC remote at the end, but only if you are using some option to save a DVC experiment, and if you don't want to push to the Git remote or DVC remote you can opt out by...).

**Comment by dberenbaum** [comment]
Closing this one since all tasks are completed. Great work @skshetry @daavoo and everyone else who helped here!

**Comment by dberenbaum** [comment]
Whoops, thanks @skshetry!
